x/datapool/keeper/pool_distribute.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	// This lastIndex is to execute the following iteration in the next block if it does not end in one block.
+	// But We're currently processing it all in one block. :)
+	lastIndex := 0
+	// TODO We need to think about how to deal with the pool where distribution has failed.
+	for i, poolID := range instantRevenueDistribute.PoolIds {
+		lastIndex = i
+		// search current pool info
+		pool, err := k.GetPool(ctx, poolID)
+		if err != nil {
+			return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+		}
+
+		// send deposit to curator
+		err = k.sendDepositToCurator(ctx, pool)
+		if err != nil {
+			return err
+		}
+
+		// distribute of revenue to each seller
+		err = k.executeRevenueDistribute(ctx, pool)
+		if err != nil {
+			return err
+		}
+	}
+	// delete this pool from instantRevenueDistribute
+	instantRevenueDistribute.RemovePreviousIndex(lastIndex)
+	k.SetInstantRevenueDistribute(ctx, instantRevenueDistribute)
+
+	return nil
+}
+
+// sendDepositToCurator returns to the Curator if the pool status is ACTIVE but the deposit has not yet been returned.
+func (k Keeper) sendDepositToCurator(ctx sdk.Context, pool *types.Pool) error {
+	if pool.Status != types.ACTIVE || pool.IsPaidDeposit {
+		return nil
+	}
+
+	curatorAddr, err := sdk.AccAddressFromBech32(pool.Curator)
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	poolAddr, err := sdk.AccAddressFromBech32(pool.PoolAddress)
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	err = k.bankKeeper.SendCoins(ctx, poolAddr, curatorAddr, sdk.NewCoins(pool.Deposit))
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	pool.IsPaidDeposit = true
+	k.SetPool(ctx, pool)
+
+	return nil
+}
+
+func (k Keeper) getAvailablePoolCoinAmount(ctx sdk.Context, pool *types.Pool) (*sdk.Int, error) {
+	poolAddr, err := sdk.AccAddressFromBech32(pool.PoolAddress)
+	if err != nil {
+		return nil, sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+	poolBalance := k.bankKeeper.GetBalance(ctx, poolAddr, assets.MicroMedDenom)
+	if pool.IsPaidDeposit {
+		return &poolBalance.Amount, nil
+	} else {
+		amount := poolBalance.Amount.Sub(pool.Deposit.Amount)
+		return &amount, nil
+	}
+}
+
+func (k Keeper) executeRevenueDistribute(ctx sdk.Context, pool *types.Pool) error {
+	availablePoolCoinAmount, err := k.getAvailablePoolCoinAmount(ctx, pool)
+	if err != nil {
+		return err
+	}
+
+	// calculate the revenue to be sent to each seller
+	eachDistributionAmount := k.getEachDistributionAmount(pool)
+
+	poolAddress, err := sdk.AccAddressFromBech32(pool.GetPoolAddress())
+	if err != nil {
+		return sdkerrors.Wrap(types.ErrRevenueDistribute, err.Error())
+	}
+
+	salesHistories := k.ListSalesHistory(ctx, pool.PoolId, pool.Round)
+	for _, history := range salesHistories {
+		// if there is no coin available, proceed no further.
+		if availablePoolCoinAmount.Equal(sdk.NewInt(0)) {
+			break
+		}
```

Comment:
This check could be done out of the `salesHistories` loop, maybe after getting `availablePoolCoinAmount`.

Reply:
Oh, Thank you.
I used the previously developed check as it is.
Your comments look good.
I fixed it. (https://github.com/medibloc/panacea-core/pull/318/commits/da195915ca26f7e51121dc3cc674a287c06dcbdb)